client/ayon_shotgrid/plugins/publish/collect_shotgrid_entities.py

```diff
@@ -0,0 +1,209 @@
+        ayon_project = context.data["projectEntity"]
+        self.log.info(ayon_project)
+
+        project_name = context.data["projectName"]
+        self.log.info(project_name)
+
+        sg_project_id = ayon_project.get("data", {}).get("shotgridId")
+
+        if not sg_project_id:
+            raise ValueError("Ayon is missing the Shotgrid project ID.")
+
+        context.data["shotgridProject"] = sg_session.find_one(
+            "Project",
+            filters=[["id", "is", int(sg_project_id)]]
+        )
+
+        self.log.info(
+            "Collected correspondig shotgrid project : {}".format(
+                context.data["shotgridProject"]
+            )
+        )
+
+        sg_assets_by_id = self._get_sg_assets_by_id(context)
+        sg_tasks_by_id = self._get_sg_tasks_by_id(context)
+
+        for instance in context:
+            instance.data["shotgridProject"] = context.data["shotgridProject"]
+
+            ayon_asset = instance.data.get("assetEntity")
+            sg_asset_id = ayon_asset.get("data", {}).get("shotgridId")
+
+            sg_asset = None
+
+            if sg_asset_id:
+                sg_asset = sg_assets_by_id.get(int(sg_asset_id))
+
+            instance.data["shotgridEntity"] = sg_asset
+            self.log.info(
+                "Collected Instance Shotgrid Asset: {}".format(sg_asset)
+            )
+
+            sg_task = None
+            ayon_task_name = instance.data.get("task")
+
+            if ayon_task_name:
+                ayon_task = self._get_ayon_task(
+                    context.data["projectName"],
+                    ayon_task_name,
+                    ayon_asset["_id"]
+                )
+
+                if ayon_task:
+                    sg_task_id = ayon_task.get("attrib", {}).get("shotgridId")
+
+                    if sg_task_id:
+                        sg_task = sg_tasks_by_id.get(
+                            int(sg_task_id)
+                        )
+
+            instance.data["shotgridTask"] = sg_task
+            self.log.info(
+                "Collected Instance Shotgrid Task: {}".format(sg_task)
+            )
+
+    def _get_sg_tasks_by_id(self, context):
+        """ Get all instances tasks from Shotgrid.
+
+        We loop over all the instances and gather all the tasks so we can
+        perform one Shotgird query and get all Tasks, then build a dict by their
+        IDs.
+
+        Args:
+            context (pyblish.Plugin.context): The current publish context.
+
+        Returns:
+            sg_tasks_by_id (dict): All the found tasks, with their ID as key
+                and the Shotgrid entity dict as value.
+        """
+        sg_session = context.data["shotgridSession"]
+        sg_tasks_ids = []
+        sg_tasks_by_id = {}
+
+        ayon_folder_ids = [
+            instance.data.get("assetEntity").get("_id")
+            for instance in context
+            if instance.data.get("assetEntity").get("_id")
+        ]
```

Comment:
```suggestion
        ayon_folder_ids = {
            instance.data["assetEntity"]["_id"]
            for instance in context
            if instance.data.get("assetEntity", {})
        }
```

Reply:
Why a `dict` instead of a `list`?

Reply:
It creates a `set()` instead of a `dict`